File: plugin-openclaw/src/params.ts
/** Paywall URL helpers — aligned with @circle-plugins/plugin-eliza params. */

export function defaultPaywallUrl(): string {
  return (
    process.env.X402_PAYWALL_URL ??
    `http://localhost:${process.env.X402_PAYWALL_PORT ?? "4021"}/risk-profile`
  );
}

/** Ensure paywall requests always hit the /risk-profile route when appropriate. */
export function normalizePaywallUrl(url: string): string {
  const trimmed = url.replace(/[),.]+$/, "");
  if (trimmed.includes("api.example.com")) {
    return defaultPaywallUrl();
  }
  if (/localhost(?::\d+)?\/?$/i.test(trimmed) || /127\.0\.0\.1(?::\d+)?\/?$/i.test(trimmed)) {
    return trimmed.replace(/\/?$/, "/risk-profile");
  }
  if (trimmed.endsWith("/risk-profile")) {
    return trimmed;
  }
  if (/localhost:\d+\/[^/]+$/i.test(trimmed) && !trimmed.endsWith("/risk-profile")) {
    return defaultPaywallUrl();
  }
  return trimmed;
}

export function resolvePaywallUrl(url?: string): string {
  return normalizePaywallUrl(url?.trim() || defaultPaywallUrl());
}

File: hermes/circle-plugins/params.py
"""Paywall URL helpers — aligned with @circle-plugins/plugin-eliza."""

from __future__ import annotations

import os
import re


def default_paywall_url() -> str:
    return os.getenv(
        "X402_PAYWALL_URL",
        f"http://localhost:{os.getenv('X402_PAYWALL_PORT', '4021')}/risk-profile",
    )


def normalize_paywall_url(url: str) -> str:
    trimmed = re.sub(r"[),.]+$", "", url)
    if "api.example.com" in trimmed:
        return default_paywall_url()
    if re.search(r"localhost(?::\d+)?/?$", trimmed, re.I) or re.search(
        r"127\.0\.0\.1(?::\d+)?/?$", trimmed, re.I
    ):
        return re.sub(r"/?$", "/risk-profile", trimmed)
    if trimmed.endswith("/risk-profile"):
        return trimmed
    if re.search(r"localhost:\d+/[^/]+$", trimmed, re.I) and not trimmed.endswith("/risk-profile"):
        return default_paywall_url()
    return trimmed


def resolve_paywall_url(url: str | None = None) -> str:
    return normalize_paywall_url((url or "").strip() or default_paywall_url())
